llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp

```diff
@@ -207,36 +215,111 @@ void InterBlockScheduling::leaveFunction() {
 }
 
 void InterBlockScheduling::enterBlock(MachineBasicBlock *BB) {
-  CurrentBlock = &getBlockState(BB);
-  CurrentBlock->resetRegion();
+  CurrentBlockState = &getBlockState(BB);
+  CurrentBlockState->resetRegion();
   DEBUG_BLOCKS(dbgs() << "  >> enterBlock " << BB->getNumber() << " "
-                      << CurrentBlock->kindAsString() << " FixPointIter="
-                      << CurrentBlock->FixPoint.NumIters << "\n");
-}
+                      << CurrentBlockState->kindAsString() << " FixPointIter="
+                      << CurrentBlockState->FixPoint.NumIters
+                      << " II=" << CurrentBlockState->FixPoint.II);
+}
+namespace {
+class PipelineExtractor : public PipelineScheduleVisitor {
+  BlockState &Loop;
+  BlockState *Prologue = nullptr;
+  BlockState *Epilogue = nullptr;
+  MachineBundle CurrentBundle;
+  std::vector<MachineBundle> TimedRegion;
+  bool InLoop = false;
+
+  void startPrologue() override {
+    // Nothing at this time, but let's keep the override around
+  }
+  void startLoop() override {
+    Prologue->BottomInsert = TimedRegion;
+    TimedRegion.clear();
+    InLoop = true;
+  }
+  void startEpilogue() override {
+    Loop.getTop().Bundles = TimedRegion;
+    TimedRegion.clear();
+    InLoop = false;
+  }
+  void finish() override {
+    Epilogue->TopInsert = TimedRegion;
+    TimedRegion.clear();
+  }
+  void startCycle() override { CurrentBundle.clear(); }
+  void emit(MachineInstr *MI) override {
+    // We re-emit the original instructions into the loop body.
+    // Prologue and epilogue obtain copies.
+    MachineInstr *ToBeEmitted =
+        InLoop ? MI : Loop.TheBlock->getParent()->CloneMachineInstr(MI);
+    CurrentBundle.add(ToBeEmitted);
+  }
+  void endCycle() override { TimedRegion.emplace_back(CurrentBundle); }
+
+public:
+  PipelineExtractor(InterBlockScheduling &InterBlock, BlockState &BS,
+                    const AIEBaseInstrInfo &TII)
+      : Loop(BS), CurrentBundle(TII.getFormatInterface()) {
+    MachineBasicBlock *LoopBlock = Loop.TheBlock;
+    for (auto *P : LoopBlock->predecessors()) {
+      if (P == LoopBlock) {
+        continue;
+      }
+      Prologue = &InterBlock.getBlockState(P);
```

Comment:
Checking: We never re-allocate the vector of `BlockStates`?

Reply:
No, only on first creation